src/check-nsfw.ts

```diff
@@ -1,192 +1,397 @@
-import path from 'node:path';
 import got from 'got';
-import { node as tensorflow } from '@tensorflow/tfjs-node';
-import { load as loadModel } from 'nsfwjs';
+import sharp from 'sharp';
+import phash from 'sharp-phash';
+import dist from 'sharp-phash/distance';
+import * as tf from '@tensorflow/tfjs-node';
+import * as nsfw from 'nsfwjs';
 import decodeGif from 'decode-gif';
-import jpeg from 'jpeg-js';
-import { AttachmentBuilder, ChannelType, EmbedBuilder, TextChannel } from 'discord.js';
+import sequelize from 'sequelize';
+import { ActionRowBuilder, AttachmentBuilder, ButtonStyle, ChannelType, EmbedBuilder, TextChannel } from 'discord.js';
 import { getDB } from '@/db';
-import config from '@/config.json';
+import { NsfwWarning } from '@/models/nsfwWarnings';
+import { notifyUser } from '@/notifications';
+import { getChannelFromSettings } from '@/util';
+import { NsfwExemption } from '@/models/nsfwExemptions';
+import { ButtonBuilder } from 'discord.js';
+import type { ButtonInteraction } from 'discord.js';
 import type { Tensor3D } from '@tensorflow/tfjs-node';
 import type { Message } from 'discord.js';
 import type { NSFWJS, predictionType } from 'nsfwjs';
 
 let model: NSFWJS;
 const GIF_MAGIC = Buffer.from([0x47, 0x49, 0x46, 0x38]);
+const NSFW_CLASSIFICATIONS = ['Porn', 'Hentai', 'Sexy'];
+
+export const ADD_NSFW_EXEMPTION = 'add-nsfw-exemption';
+export const REMOVE_NSFW_EXEMPTION = 'remove-nsfw-exemption';
+
+export async function loadModel(): Promise<void> {
+	if (!model) {
+		tf.enableProdMode();
+		model = await nsfw.load('InceptionV3');
+	}
+}
 
 export async function checkNSFW(message: Message, urls: string[]): Promise<void> {
-	if (message.channel instanceof TextChannel && message.channel.nsfw) {
+	const nsfwDetectionEnabled = getDB().get('nsfw.enabled') === 'true';
+
+	if (!nsfwDetectionEnabled || (message.channel instanceof TextChannel && message.channel.nsfw)) {
 		return; // * Do not check if the channel is NSFW
 	}
 
-	if (!model) {
-		const quantized = config.quantized_nsfw_model ? 'quantized' : 'normal';
-		const modelPath = path.join(path.resolve(config.nsfw_model_path), quantized);
-		const modelUri = `file://${modelPath}/`;
-		
-		// * We have to set the options to `any` here because although `type` is a valid option
-		// * it's not included in the types, this is also an inline type so we can't enhance it
-		model = await loadModel(modelUri, { type: 'graph' } as any);
-	}
+	let exemptionDistance: number = 0;
+	try {
+		exemptionDistance = parseInt(getDB().get('nsfw.exemption.distance') ?? '0');
+	} catch {}
```

Comment:
Aha, I was thinking with my java head on that this would throw an exception if parsing failed (if someone tried to set this to a value which isn't a whole number) but it just returns `NaN` instead, I'll fix that up, there are a couple of places where I'm doing this